### DISCUSSION TITLE
Make cancelling deployments easier

### DISCUSSION THREAD

**Discussion by RobSchilderr** [vercel/community#4280] [Feature Question]
### Question

In scenarios where a Turborepo hosts multiple projects, pushing changes to Vercel followed by rapid merging from feature branches to staging, and subsequently to production, can trigger a flurry of deployments. Manually cancelling these deployments is a time-consuming task, especially when dealing with a high number of concurrent deployments.

Initially, I considered leveraging the Vercel CLI to automate the cancellation process. However, I encountered several issues that hindered progress. Therefore, I am proposing an enhancement to simplify the deployment cancellation process, particularly when interacting with GitHub.

Here's a glimpse of the issues faced:

I am facing issues with my Vercel token authentication while trying to cancel deployments using a script. For example, I have a token in my personal account with the scope set to the "organisation" team, named "organisation-token-new". Despite setting the correct scope, I am unable to perform certain actions like cancelling deployments. The error message I receive is:

`Not authorized: Trying to access resource under scope 'privateaccount'. You must re-authenticate to this scope or use a token with access to this scope.`

I have already tried re-authenticating and generating a new token, but the issue persists. I have also ensured that the token is being used correctly in the Authorization header of my API requests.

Below is the script I am attempting to run:

```#!/bin/bash

# Load the Vercel token from the environment variable
source /.env

# Set your team ID
TEAM_ID=""

# List deployments
DEPLOYMENTS=$(curl "https://api.vercel.com/v6/now/deployments?teamId=$TEAM_ID" \
  -H "Authorization: Bearer $VERCEL_TOKEN" \
  -H "Content-Type: application/json")

# Parse and limit the deployments using jq
DEPLOYMENTS=$(echo "$DEPLOYMENTS" | jq '.deployments | .[:5]')

# Filter the deployments to get only the ones created by 'x@gmail.com'
USER_DEPLOYMENTS=$(echo "$DEPLOYMENTS" | jq '.[] | select(.creator.email == "x@gmail.com")')

# Loop over the user's deployments
for DEPLOYMENT in $(echo "${USER_DEPLOYMENTS}" | jq -r '.uid'); do
  # Cancel the deployment
  RESPONSE=$(curl -X PATCH "https://api.vercel.com/v12/now/deployments/$DEPLOYMENT/cancel" \
    -H "Authorization: Bearer $VERCEL_TOKEN" \
    -H "Content-Type: application/json")
  
  # Check if the cancellation was successful
  if [ "$(echo "$RESPONSE" | jq -r '.error')" == "null" ]; then
    echo "Successfully cancelled deployment with ID: $DEPLOYMENT"
  else
    echo "Failed to cancel deployment with ID: $DEPLOYMENT"
    echo "Error message: $(echo "$RESPONSE" | jq -r '.error.message')"
  fi
done
```

My ultimate goal is to use the CLI and cancel deployments, but these issues arise. I am heavily invested in the GitHub - Vercel integration.

I would appreciate any assistance or insights into how other people are dealing with this issue. Besides I would also appreciate assistance into why my token is not working as expected and how I can resolve this issue to successfully cancel deployments using my script.

Thank you.

PS: With the Vercel team I opened a case for support, however after 10 days I have received no reply, hence I decide to make this a public discussion to see if other people also have this problem.

**Comment by amyegan** [comment]
Thanks for reaching out, @RobSchilderr! I was only able to reproduce the auth problem you described if I used an invalid token or used an incompatible `teamId` value. The value can be either the team slug (e.g. `my-team-name`) or the team identifier (e.g. `team_1a2b3c4d5e...`) which can be found on the team's settings page. Mismatched team id is a fairly common cause of access issues, so I'd recommend double checking that.

I also have an alternative approach that might work better for you. Instead of trying to cancel extra builds, you could use [Ignored Build Step](https://vercel.com/docs/projects/overview#ignored-build-step) to block unnecessary builds from starting at all. You could use the `Only build Turborepo app if there are changes` option or customize your own script to handle it however you like. 

Another option is to use GitHub Actions instead of the [Vercel for GitHub](https://vercel.com/docs/deployments/git/vercel-for-github) automation. You could trigger the deployments on [`release`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release) or [`workflow_dispatch`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch) instead of on `push`. That would give you more control over when they should happen. A guide covering setup of a basic action script can be found at https://vercel.com/guides/how-can-i-use-github-actions-with-vercel
